engine/consensus/compliance/engine.go

```diff
@@ -149,9 +173,108 @@ func (e *Engine) ProcessLocal(event interface{}) error {
 // Process processes the given event from the node with the given origin ID in
 // a blocking manner. It returns the potential processing error when done.
 func (e *Engine) Process(originID flow.Identifier, event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(originID, event)
-	})
+	e.pendingEventSink <- &Event{
+		OriginID: originID,
+		Msg:      event,
+	}
+	return nil
+}
+
+// processEvents is processor of pending events which drives events from networking layer to business logic in `Core`.
+// Effectively consumes messages from networking layer and dispatches them into corresponding sinks which are connected with `Core`.
+// Should be run as a separate goroutine.
+func (e *Engine) processEvents() {
+	// takes pending event from one of the queues
+	// nil sink means nothing to send, this prevents blocking on select
+	fetchEvent := func() (*Event, EventSink, *fifoqueue.FifoQueue) {
+		if val, ok := e.pendingBlocks.Front(); ok {
+			return val.(*Event), e.blockSink, e.pendingBlocks
+		}
+		if val, ok := e.pendingVotes.Front(); ok {
+			return val.(*Event), e.voteSink, e.pendingVotes
+		}
+		return nil, nil, nil
+	}
+
+	for {
+		pendingEvent, sink, fifo := fetchEvent()
+		select {
+		case event := <-e.pendingEventSink:
+			e.processPendingEvent(event)
+		case sink <- pendingEvent:
```

Comment:
what if `pendingEvent` is nil?

Reply:
that is great, means it won't try sending and won't participate in select statement

Reply:
if `pendingEvent` is nil, `sink` is nil. Sending to a nil channel blocks forever. ✔️